bionic/code_analyzer.py

```diff
@@ -0,0 +1,203 @@
+now. But in the future, we will add more functionality like checking
+function references and any changes to class objects.
+
+It is used by Bionic to detect any changes to entity functions. When
+the entity function changes, the value of the Bionic entity can change,
+which requires Bionic to invalidate the cache and compute the entity
+again.
+
+This logic is only used when versioning mode is set to "auto" or
+"assisted". Since the logic doesn't cover every type of code change,
+like references or classes, this logic is still experimental. When it
+reaches a state where it detects all kinds of changes, we will make
+this the default behavior.
+"""
+
+import inspect
+import json
+import warnings
+
+from .utils.misc import oneline
+
+
+# We use object names for circular references when analyzing code. For
+# any objects that don't have a name, we use this randomly generated
+# default value.
+DEFAULT_NAME = b"zA)T+ndu/Ygd%JH=eux)[F9:<>)zND]}u4x,c49tWH<yd5+#+4Xg{=J<[S5/'XG+"
+
+
+def analyze_func(func):
+    assert inspect.isroutine(func)
+
+    analyzer = CodeAnalyzer()
+    analyzer.analyze(func)
+    return analyzer.get_analyzed_bytes()
+
+
+class CodeAnalyzer:
+    """
+    This class contains the logic to analyze code objects and convert
+    them into a bytearray. For now, it only checks the function's
+    bytecode and constants. But in the future, it will be able to
+    analyze its references, including class types.
+    """
+
+    def __init__(self):
+        self._results_by_key = {}
+        self._keys_being_hashed = set()
+        self._analyzed_bytearray = bytearray()
+
+    def get_analyzed_bytes(self):
+        return self._analyzed_bytearray
+
+    def analyze(self, obj):
+        """
+        Analyzes the object and adds the converted bytecode to the
+        result. It caches the objects to avoid duplicate work and also
+        handles circular references by hashing their names instead.
+        """
+        obj_key = get_obj_key(obj)
+
+        # If already hashed, use the hash.
+        if obj_key in self._results_by_key:
+            self._add_results(obj, self._results_by_key[obj_key])
+            return
+
+        # If the obj is already hashing, break the circular ref by
+        # analyzing the name instead.
+        if obj_key in self._keys_being_hashed:
+            name = get_obj_name(obj)
+            if name is None:
+                name = DEFAULT_NAME
+            obj = name
+            obj_key = get_obj_key(obj)
+
+        # Add a prefix for obj type so we can differentiate between
+        # different types that look the same, like a list and a tuple,
+        # or an int and bytes.
+        self._add_prefix_for_type(obj)
+
+        self._keys_being_hashed.add(obj_key)
+        self._analyze(obj)
+        self._keys_being_hashed.remove(obj_key)
+
+    def _add_prefix_for_type(self, obj):
+        if isinstance(obj, bytes):
+            prefix = b"A"
```

Comment:
I did that too for this one, but then I used `B` for both bytes and bool. But if we move them with their encoding logic, I think we should have some mnemonic relationship. Maybe I can change it to 2 bytes instead of 1. Wdyt?